### PR TITLE
fix: multiaddr address handling

### LINKED ISSUE #6: Fix multiaddr address handling
https://github.com/ipfs-shipyard/ipfs-redux-bundle/blob/82a6e030d3022732b7efde2fba6775e70d7d3483/index.js#L135 assumes the host is ip4. but it could be ip6.

https://github.com/ipfs-shipyard/ipfs-redux-bundle/blob/82a6e030d3022732b7efde2fba6775e70d7d3483/index.js#L167 assumes http but the mutliaddr may specify https.

### PR DESCRIPTION
**Pull request by hacdias**
Closes #6.

License: MIT
Signed-off-by: Henrique Dias <hacdias@gmail.com>

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -159,13 +164,18 @@ module.exports = (opts = {}) => {
...
-      addr = addr.split('/')
-
-      store.doUpdateIpfsApiOpts({
-        host: addr[2],
-        port: addr[4],
-        protocol: 'http'
-      })
+      try {
+        const maddr = multiaddr(addr).nodeAddress()
+        const https = addr.includes('https')
```
Comment (olizilla) [inline index.js]: addr.includes is easily fooled by dnsaddrs like `/dnsaddr/httpseverywhere/tcp/80`

File: index.js [hunk 2]
```diff
@@ -132,7 +133,11 @@ module.exports = (opts = {}) => {
 
     selectIpfsApiOpts: state => state.ipfs.apiOpts,
 
-    selectIpfsApiAddress: state => `/ip4/${state.ipfs.apiOpts.host}/tcp/${state.ipfs.apiOpts.port}`,
+    selectIpfsApiAddress: state => {
+      const ipv = state.ipfs.apiOpts.host.split('.').length === 4 ? 'ip4' : 'ip6'
+
+      return `/${ipv}/${state.ipfs.apiOpts.host}/tcp/${state.ipfs.apiOpts.port}`
```
Comment (olizilla) [inline index.js]: if the user sets a domain name as the host then the multiaddr should start with `/dnsaddr` not `/ip*`

It'd be best to use  http://multiformats.github.io/js-multiaddr/ and pick out the info from there. If the api for that is awkward than we can improve it.

**Comment by olizilla** [comment]
@hacdias please add test cases for `ip4`, `ip6`, `dnsaddr` `dns4`, `dns6` in `http` and `https` flavour

**Comment by hacdias** [comment]
@olizilla I made some changes: now it saves the options in the reliable format possible: if you set an options object it will save it as an options object, if you set an address, it will save a multiaddress object. This way we'll never lose information and

**Comment by hacdias** [comment]
So I don't believe we now need those tests since those should be taken care of in js-ipfs-api.

**Comment by olizilla** [comment]
An api that could return an options object _or_ a multiaddr pushes the complexity on the calling code.

Can we simplify this? Right now we need to support letting the user set the api address as a multiaddr. This is (relatively) user friendly as when you start your daemon it prints out the api address in mutliaddr format, so you can just copy and paste. When you update your config file you have to specify it the api in multiaddr format too, so again, works nicely for copy and paste.

When the api address is updated we store it in localStorage. We don't need to tell the user about it, as webui provides a UI to do that. So we don't need to deal with an api opts object coming from localStorage right now. **Let's always deal with ipfsApi as a multiaddr.** And lets only provide selectors to get and set it as a multiaddr.

As for jsipfs, it's kinda nice to let users customise their entire options config, so that should still be an options object. 

I'm happy to keep the js-ipfs logic as it is, but it's worth considering that it's an odd situation to let the calling code decide what version of js-ipfs to use but not what options to initiate it with... and on the flip side to not let the user decide which version of js-ipfs to use but to let them pick the options. I don't think that's a huge problem for us right now, but we should think about a better solution for a future PR. Either the app should get to choose the opts (it could still ask the user) or the user should be able to pick an ipfs verion. That would require us to publish ipfs release to IPFS, so we could lazy load them at runtime, which would be nice, but isn't a thing yet.

**Comment by hacdias** [comment]
@olizilla now we just pass addresses around. I kept the `ipfsOpts` in localStorage as a string because there might be other packages that could take advantage of this.

File: index.js [hunk 3]
```diff
@@ -205,7 +191,7 @@ async function tryApi (opts) {
     console.time('js-ipfs-api ready!')
     console.log('Trying ipfs-api', opts)
 
-    console.info('🎛️ Customise your js-ipfs-api options by storing a `ipfsApi` object in localStorage. e.g. localStorage.setItem(\'ipfsApi\', JSON.stringify({port: \'1337\'}))')
+    console.info('🎛️ Customise your js-ipfs-api options by storing a `ipfsApi` object in localStorage. e.g. localStorage.setItem(\'ipfsApi\', JSON.stringify(\'/ip4/127.0.0.1/tcp/5001\'))')
```
Comment (olizilla) [inline index.js]: no need to JSON.stringify a string!

File: index.js [hunk 4]
```diff
@@ -8,18 +8,14 @@ const defaultOptions = {
...
-  }
+  defaultApiAddress: '/ip4/127.0.0.1/tcp/5001'
 }
 
 module.exports = (opts = {}) => {
   opts = Object.assign({}, defaultOptions, opts)
 
   const defaultState = {
-    apiOpts: opts.defaultApiOpts,
+    apiAddress: opts.defaultApiAddress,
```
Comment (olizilla) [inline index.js]: let's init this to the localStorage value if available

File: index.js [hunk 5]
```diff
@@ -63,15 +59,17 @@ module.exports = (opts = {}) => {
...
     if (opts.tryApi) {
-      let apiOpts = getState().ipfs.apiOpts
+      let apiAddress = getState().ipfs.apiAddress
+      let userOpts = getUserOpts('ipfsApi')
 
-      if (typeof apiOpts === 'object') {
-        apiOpts = Object.assign({}, apiOpts, getUserOpts('ipfsApi'))
+      if (userOpts !== apiAddress && typeof userOpts === 'string') {
+        apiAddress = userOpts
+        dispatch({ type: 'IPFS_API_OPTS_UPDATED', payload: userOpts })
```
Comment (olizilla) [inline index.js]: 👍

File: index.js (hunk 4 above)
Comment (hacdias) [reply index.js]: Done"

**Comment by hacdias** [comment]
@olizilla I believe it's finally done!

**Comment by olizilla** [comment]
As we're not using the multiaddr provided by the user if we find it to be invalid we're going to need to inform them about it.

Create a helper function that grabs the ipfsApi from localStorage, validates it, an either logs an error if its invalid, or returns a valid multiaddr

**Comment by olizilla** [comment]
Where the user sets a multiaddr via the `do` function, they'd want to know straight away if they've set an incorrect one, so we're going to need to add actions and a state for invalid multiaddrs

**Comment by olizilla** [comment]
Have a think about what would be the ideal user and dev experience for using this!

**Comment by olizilla** [comment]
I'm gonna merge this and create an issue for alerting the user about invalid multiaddrs

**Pull request merged**